### PR TITLE
Dev/firebase futures

### PR DESCRIPTION
**Pull request by TheBlueHawk**
Moving the Firestore interactions from void methods with callbacks to CompletableFutures methods

### CODE REVIEW

**Review by AdrienRem — APPROVED** [review]
It took me a solid 45 minutes to read but I don't have much to say. Good job on the occasional refactor (like in the HeatMapHandler) and good idea to run the autoformatter on some code files.

Maybe the refactor should have been an other PR but since you were already modifying the files I can't blame you because you're improving the code a bit.

File: app/src/androidTest/java/ch/epfl/sdp/contamination/GridSenderTest.java [hunk 1]
```diff
@@ -103,108 +93,81 @@ public void setupMockito() {
...
-
-    class MockGridInteractor extends GridFirestoreInteractor {
+        // Get reference to UI handler
+        Handler uiHandler = mActivityRule.getActivity().uiHandler;
 
-        // TODO: GridFirestoreInteractor should become an interface too
-        MockGridInteractor() {
-            super();
-        }
+        writeSuccessToUi = (a) -> uiHandler.post(() -> exchangeStatus.setText("EXCHANGE Succeeded"));
```
Comment (Ulydev) [inline app/src/androidTest/java/ch/epfl/sdp/contamination/GridSenderTest.java:102]: Would be nice to move these strings to strings.xml

File: app/src/main/java/ch/epfl/sdp/HistoryFirestoreInteractor.java [hunk 2]
```diff
@@ -1,42 +1,35 @@
...
         String path = "History/" + user.getId() + "/Positions";
-        fsi.readCollection(path, handler);
+        return readCollection(collectionReference(path));
     }
 
 
-    public void write(Map<String, Object> content, OnSuccessListener success,
-                      OnFailureListener failure) {
+    public CompletableFuture<Void> write(Map<String, Object> content) {
         String path = "History/" + user.getId() + "/Positions";
```
Comment (Ulydev) [inline app/src/main/java/ch/epfl/sdp/HistoryFirestoreInteractor.java]: Would be nice to move the path resolution to a separate function to reuse it or change it later :-)

File: app/src/main/java/ch/epfl/sdp/HistoryFragment.java [hunk 3]
```diff
@@ -57,44 +61,31 @@ public void onCreate(Bundle savedInstanceState) {
...
+            ListView historyTracker = view.findViewById(R.id.history_tracker);
+            historyTracker.setAdapter(historyAdapter);
+            historyAdapter.clear();
+
+            connectionStatus.setText(R.string.query_ok);
+
+            for (Map.Entry<String, Map<String, Object>> mapEntry : res.entrySet()) {
+                try {
+                    Map<String, Object> positionRecord = mapEntry.getValue();
+                    historyAdapter.insert(String.format("Found @ %s:%s on %s",
```
Comment (Ulydev) [inline app/src/main/java/ch/epfl/sdp/HistoryFragment.java:78]: We can use resource strings with interpolation to remove this hardcoded string, but that is out of the scope for this PR

File: app/src/main/java/ch/epfl/sdp/Map/HeatMapHandler.java [hunk 4]
```diff
@@ -5,70 +5,77 @@
...
         this.map = map;
-        initFireBaseQueryHandler();
+        initQuery();
     }
 
+    private void initQuery() {
+        db.readCollection(collectionReference("LastPositions"))
+                .thenAccept(value -> createGeoJson(value))
+                .exceptionally(e -> {
+                    Toast.makeText(parentClass.getActivity(), "Cannot retrieve " +
```
Comment (Ulydev) [inline app/src/main/java/ch/epfl/sdp/Map/HeatMapHandler.java:63]: Same here for string interpolation

File: app/src/main/java/ch/epfl/sdp/contamination/ConcreteAnalysis.java [hunk 5]
```diff
@@ -102,27 +101,28 @@ private void modelInfectionEvolution(Map<Carrier, Integer> suspectedContacts) {
...
     @Override
-    public void updateInfectionPredictions(Location location, Date startTime, Callback<Void> callback) {
-
+    public CompletableFuture<Void> updateInfectionPredictions(Location location, Date startTime) {
         Date now = new Date(System.currentTimeMillis());
-        receiver.getUserNearbyDuring(location, startTime, now, aroundMe -> {
-            modelInfectionEvolution(identifySuspectContacts(aroundMe));
-            callback.onCallback(null);
-        });
+        CompletableFuture future1 = receiver.getUserNearbyDuring(location, startTime, now)
```
Comment (Ulydev) [inline app/src/main/java/ch/epfl/sdp/contamination/ConcreteAnalysis.java]: Would be useful for maintenance to give more explicit names for futures, or even directly inline them into CompletableFuture.allOf?

File: app/src/main/java/ch/epfl/sdp/contamination/InfectionFragment.java [hunk 6]
```diff
@@ -84,18 +92,18 @@ public void onModelRefresh(View v) {
...
-                    infectionStatus.setText(analyst.getCarrier().getInfectionStatus().toString());
-                    infectionProbability.setProgress(Math.round(analyst.getCarrier().getIllnessProbability() * 100));
-                    Log.e("PROB:", analyst.getCarrier().getIllnessProbability() + "");
-                });
-            });
-        });
+        service.getReceiver().getMyLastLocation(AccountFragment.getAccount(getActivity()))
+                .thenApply(location -> service.getAnalyst().updateInfectionPredictions(location, refreshTime)
+                        .thenAccept(n -> {
+                            infectionStatus.setText("Posted!!");
```
Comment (Ulydev) [inline app/src/main/java/ch/epfl/sdp/contamination/InfectionFragment.java]: Resource string

**Review by Ulydev — APPROVED** [review]
An overall great work and refactoring using more modern callback handling! Makes for a game-changer PR 👍👍

File: app/src/main/java/ch/epfl/sdp/HistoryFirestoreInteractor.java (hunk 2 above)
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/HistoryFirestoreInteractor.java]: Good point, I'm going to do that

File: app/src/main/java/ch/epfl/sdp/HistoryFragment.java (hunk 3 above)
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/HistoryFragment.java:78]: I left a TODO for next PR

File: app/src/main/java/ch/epfl/sdp/Map/HeatMapHandler.java (hunk 4 above)
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/Map/HeatMapHandler.java:63]: I left a TODO for next PR

File: app/src/main/java/ch/epfl/sdp/contamination/ConcreteAnalysis.java (hunk 5 above)
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/contamination/ConcreteAnalysis.java]: Refactored

File: app/src/main/java/ch/epfl/sdp/contamination/InfectionFragment.java (hunk 6 above)
Comment (TheBlueHawk) [reply app/src/main/java/ch/epfl/sdp/contamination/InfectionFragment.java]: Done

**Pull request merged**